apps/gdalalg_dispatcher.h

```diff
@@ -0,0 +1,242 @@
+/******************************************************************************
+ *
+ * Project:  GDAL
+ * Purpose:  gdal subcommand dispatcher
+ * Author:   Even Rouault <even dot rouault at spatialys.com>
+ *
+ ******************************************************************************
+ * Copyright (c) 2024, Even Rouault <even dot rouault at spatialys.com>
+ *
+ * SPDX-License-Identifier: MIT
+ ****************************************************************************/
+
+#ifndef GDALALG_DISPATCHER_INCLUDED
+#define GDALALG_DISPATCHER_INCLUDED
+
+#include "gdalalgorithm.h"
+
+#include "gdal_priv.h"
+
+//! @cond Doxygen_Suppress
+
+/************************************************************************/
+/*                       GDALDispatcherAlgorithm                        */
+/************************************************************************/
+
+template <class RasterDispatcher, class VectorDispatcher>
+class GDALDispatcherAlgorithm : public GDALAlgorithm
+{
+  public:
+    GDALDispatcherAlgorithm(const std::string &name,
+                            const std::string &description,
+                            const std::string &helpURL)
+        : GDALAlgorithm(name, description, helpURL),
+          m_rasterDispatcher(std::make_unique<RasterDispatcher>(
+              /* openForMixedRasterVector = */ true)),
+          m_vectorDispatcher(std::make_unique<VectorDispatcher>())
+    {
+        // A "info" dispacher command is a shortcut for something like
+        // "raster info", "vector info". Best to expose the latter.
+        SetDisplayInJSONUsage(false);
+    }
+
+    bool
+    ParseCommandLineArguments(const std::vector<std::string> &args) override;
+
+    std::string GetUsageForCLI(bool shortUsage,
+                               const UsageOptions &usageOptions) const override;
+
+  private:
+    std::unique_ptr<RasterDispatcher> m_rasterDispatcher{};
+    std::unique_ptr<VectorDispatcher> m_vectorDispatcher{};
+    bool m_showUsage = true;
+
+    bool RunImpl(GDALProgressFunc, void *) override
+    {
+        CPLError(CE_Failure, CPLE_AppDefined,
+                 "The Run() method should not be called directly on the \"gdal "
+                 "%s\" program.",
+                 GetName().c_str());
+        return false;
+    }
+};
+
+/************************************************************************/
+/*         GDALDispatcherAlgorithm::ParseCommandLineArguments()         */
+/************************************************************************/
+
+template <class RasterDispatcher, class VectorDispatcher>
+bool GDALDispatcherAlgorithm<RasterDispatcher, VectorDispatcher>::
+    ParseCommandLineArguments(const std::vector<std::string> &args)
+{
+    bool ok;
+    {
```

Comment:
following comment added:
    // We first try to process with the raster specific algorithm (that has
    // been instantiated in a special way to accept both raster and vector
    // input datasets). If the raster specific algorithm can parse successfully
    // the arguments *and* the dataset is a raster one, then continue processing
    // with it. Otherwise try with the vector specific algorithm.